### ISSUE TITLE
Event processing approach

### ISSUE THREAD

**Issue by rusu24edward** [LLNL/Abmarl#337]
Currently, the actors process the events and manage necessary state changes. #319 shows that this is a limiting design, and it is not entirely clear how to proceed with this enhancement.

I propose the following processing flow:
1. Agents report their actions in the `action_dict` of the `step` function.
2. Actions are passed to the actors, which report the outcome as an `Event` object.
3. The simulation records the events in order (since the simulation specifies the order of the actors, then it naturally follows that the order of the events is also desired).
4. The simulation updates its state from the `events` as per its own design. State objects can handle this.

In short, this is already the design that we had in mind (see #318), but now we're making a specific object  called `Event` which manages actor outcome in an ordered fashion.

**Comment by rusu24edward** [comment]
Attack actors report success if at least one of their attacks is successful, but this doesn't capture a situation like if the attacker launches 10 attacks and only 1 of them is successful.

**Comment by rusu24edward** [comment]
There are also many different types of events that aren't really captured by an actor. For example, if an agent overlaps with a teleportation cell, then it should be teleported to a different cell. This isn't really an Actor in the typical sense because the agent doesn't choose to teleport, it chose to move onto a cell, and the state of game indicates that a teleportation should happen. Another example is when we compute deaths by overlap instead of by direct attacks. These types of events need to be captured better.

**Comment by rusu24edward** [comment]
Difference between proposed event and actual event. A proposed event comes out of the actor class as a proposed action, which the state handler take and processes. It may or may not happen the way it is proposed. For example, the attack actor says that the attack was successful, here's who attacked whom. But then the state handler can specify that the attacked agent needs to be attacked twice in the same step, which didn't happen, so it rejects the proposed events.

Actual events are the "output" from the step function. They record actual change of state. Example, the agent actually moved from here to there. We need this for determining rewards, since rewards are a function of (state, action, next_state).
